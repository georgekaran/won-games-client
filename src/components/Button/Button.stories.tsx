import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import Button from './Button'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />

Default.args = {
  children: 'Button'
}
