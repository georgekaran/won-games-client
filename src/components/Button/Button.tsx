import React from 'react'
import * as S from './Button.styles'

export type ButtonProps = {
  children?: React.ReactNode
}

const Button = ({ children }: ButtonProps) => (
  <S.Wrapper>
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
