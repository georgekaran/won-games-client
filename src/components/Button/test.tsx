import React from 'react'
import { RenderResult, screen } from '@testing-library/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button, { ButtonProps, ButtonTypes } from '.'
import { renderWithTheme } from '@/test/helpers'
import theme from '@/styles/theme'

type SutProps = Omit<ButtonProps, 'children'> & ButtonTypes

const makeSut = ({ size = 'medium', fullWidth = false, icon = null, as = 'button', minimal = false, ...props }: SutProps = {}): RenderResult => {
  return renderWithTheme(<Button size={size} fullWidth={fullWidth} icon={icon} as={as} minimal={minimal} {...props}>Button</Button>)
}

describe('<Button />', () => {
  test('should render with correct initial state', () => {
    const { container } = makeSut()
    expect(container.firstChild).toMatchSnapshot()
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '4rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.small)
    expect(button).toHaveStyleRule('padding', `${theme.spacings.xxsmall} ${theme.spacings.medium}`)
  })

  test('should render small if small size is provided', () => {
    makeSut({ size: 'small' })
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '3rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.xsmall)
    expect(button).toHaveStyleRule('padding', theme.spacings.xxsmall)
  })

  test('should render large if large size is provided', () => {
    makeSut({ size: 'large' })
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '5rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.medium)
    expect(button).toHaveStyleRule('padding', `${theme.spacings.xxsmall} ${theme.spacings.xlarge}`)
  })

  test('should render fullWidth if prop is provided', () => {
    makeSut({ fullWidth: true })
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('width', '100%')
  })

  test('should render icon if an icon is provided', () => {
    makeSut({ icon: <AddShoppingCart data-testid="icon" /> })
    const icon = screen.getByTestId('icon')
    expect(icon).toBeInTheDocument()
    expect(screen.getByText(/Button/)).toBeInTheDocument()
  })

  test('should render a minimal version', () => {
    makeSut({ icon: <AddShoppingCart data-testid="icon" />, minimal: true })
    const icon = screen.getByTestId('icon')
    expect(icon).toBeInTheDocument()
    const button = screen.getByRole('button', { name: /Button/i })
    expect(button).toHaveStyle({
      background: 'none',
      color: '#F231A5'
    })
    expect(button).toHaveStyleRule(
      'background',
      'none',
      {
        modifier: ':hover'
      }
    )
  })

  test('should render as anchor if prop \'as\' is a', () => {
    makeSut({ as: 'a', href: '/link' })
    const buttonAsLink = screen.getByRole('link', { name: /Button/i })
    expect(buttonAsLink).toBeInTheDocument()
  })
})
