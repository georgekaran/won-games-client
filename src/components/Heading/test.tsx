import React from 'react'
import { RenderResult, screen } from '@testing-library/react'

import Heading, { HeadingProps } from '.'
import { renderWithTheme } from '@/test/helpers'
import theme from '@/styles/theme'

type SutProps = Omit<HeadingProps, 'children'>

const makeSut = ({ color = 'white', lineLeft = false, lineBottom = false, size = 'medium', lineColor = 'primary' }: SutProps): RenderResult => {
  return renderWithTheme(
    <Heading
      color={color}
      lineLeft={lineLeft}
      lineBottom={lineBottom}
      size={size}
      lineColor={lineColor}>
      Random Heading
    </Heading>)
}

describe('<Heading />', () => {
  test('should render with initial state', () => {
    const { container } = makeSut({})
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toBeInTheDocument()
    expect(heading).toHaveStyleRule('color', theme.colors.white)
    expect(heading).not.toHaveStyleRule('padding-left')
    expect(heading).not.toHaveStyleRule('padding-bottom')
    expect(container.firstChild).toMatchSnapshot()
  })

  test('should render Heading with black text if color black is provided', () => {
    makeSut({ color: 'black' })
    expect(screen.getByRole('heading', { name: /Random Heading/ })).toHaveStyleRule('color', theme.colors.black)
  })

  test('should present a line in the left if lineLeft is true', () => {
    makeSut({ lineLeft: true })
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toHaveStyleRule('padding-left', theme.spacings.xxsmall)
    expect(heading).toHaveStyleRule('border-left', `0.7rem solid ${theme.colors.primary}`)
  })

  test('should present a line in the bottom if lineBottom is true', () => {
    makeSut({ lineBottom: true })
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toHaveStyleRule('margin-bottom', theme.spacings.medium)
    expect(heading).toHaveStyleRule(
      'background-color',
      `${theme.colors.primary}`, {
        modifier: '::after'
      }
    )
  })

  test('should render with small size if size prop is provided', () => {
    makeSut({ size: 'small' })
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toHaveStyle({
      'font-size': '1.6rem'
    })

    expect(heading).toHaveStyleRule(
      'width',
      '3rem', {
        modifier: '::after'
      }
    )
  })

  test('should change Heading line color to primary', () => {
    makeSut({ lineBottom: true, lineLeft: true, lineColor: 'primary' })
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toHaveStyle({
      'border-left': `0.7rem solid ${theme.colors.primary}`
    })
    expect(heading).toHaveStyleRule(
      'background-color',
      theme.colors.primary,
      {
        modifier: '::after'
      })
  })

  test('should change Heading line color to secondary', () => {
    makeSut({ lineBottom: true, lineLeft: true, lineColor: 'secondary' })
    const heading = screen.getByRole('heading', { name: /Random Heading/ })
    expect(heading).toHaveStyle({
      'border-left': `0.7rem solid ${theme.colors.secondary}`
    })
    expect(heading).toHaveStyleRule(
      'background-color',
      theme.colors.secondary,
      {
        modifier: '::after'
      })
  })
})
