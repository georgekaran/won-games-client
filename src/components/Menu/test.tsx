import React from 'react'
import faker from 'faker'
import { screen, RenderResult, fireEvent } from '@testing-library/react'

import Menu, { MenuProps } from '.'
import { renderWithTheme } from '@/test/helpers'

const makeSut = ({ username = '' }: MenuProps = {}): RenderResult => {
  return renderWithTheme(<Menu username={username} />)
}

describe('<Menu />', () => {
  test('should render with initial state', () => {
    const { container } = makeSut()
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getByLabelText(/open menu/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/search/i)).toBeInTheDocument()
    expect(screen.getByLabelText(/open shopping cart/i)).toBeInTheDocument()
    expect(screen.getByTestId('logo')).toBeInTheDocument()
  })

  test('should handle the open/close mobile menu', () => {
    makeSut()
    // Select the menu
    const menuFull = screen.getByRole('navigation', { hidden: true })
    // Check if menu is hide
    expect(menuFull.getAttribute('aria-hidden')).toBe('true')
    expect(menuFull).toHaveStyleRule('opacity', '0')
    // Click in the button to open the menu e check if the menu is open
    const menuIcon = screen.getByLabelText(/open menu/i)
    fireEvent.click(menuIcon)
    expect(menuFull.getAttribute('aria-hidden')).toBe('false')
    expect(menuFull).toHaveStyleRule('opacity', '1')
    // Click in the button to close the menu e check if the menu is hide/close
    const closeMenuIcon = screen.getByLabelText(/close menu/i)
    fireEvent.click(closeMenuIcon)
    expect(menuFull.getAttribute('aria-hidden')).toBe('true')
    expect(menuFull).toHaveStyleRule('opacity', '0')
  })

  test('Should show My account and wishlist links if username is not nullish', () => {
    makeSut({ username: faker.internet.userName() })
    expect(screen.getByText(/my account/i)).toBeInTheDocument()
    expect(screen.getByText(/wishlist/i)).toBeInTheDocument()
  })

  test('Should not show RegisterBox if username is not nullish', () => {
    makeSut({ username: faker.internet.userName() })
    expect(screen.queryByText(/Log in now/i)).not.toBeInTheDocument()
    expect(screen.queryByText(/Sign up/i)).not.toBeInTheDocument()
  })
})
