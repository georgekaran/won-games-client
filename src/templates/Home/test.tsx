import React from 'react'
import { RenderResult, screen } from '@testing-library/react'

import Home from '.'
import { renderWithTheme } from '@/test/helpers'
import bannerMock from '@/components/BannerSlider/mock'
import gameCardSliderMock from '@/components/GameCardSlider/mock'
import highlightMock from '@/components/Highlight/mock'

const homeProps = {
  banners: bannerMock,
  newGamesSlider: gameCardSliderMock,
  mostPopularHighlight: highlightMock,
  mostPopularGames: gameCardSliderMock,
  upcommingGames: gameCardSliderMock,
  upcommingGamesHighlight: highlightMock,
  upcommingMoreGames: gameCardSliderMock,
  freeGames: gameCardSliderMock,
  freeHighlight: highlightMock
}

const makeSut = (): RenderResult => {
  return renderWithTheme(<Home {...homeProps} />)
}

describe('<Home />', () => {
  it('should render menu and footer', () => {
    makeSut()
    expect(screen.getByLabelText(/open menu/i)).toBeInTheDocument()
    expect(screen.getByText(/follow us/i)).toBeInTheDocument()
  })

  it('should render heading sections', () => {
    makeSut()
    expect(screen.getByRole('heading', { name: /News/i })).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: /Most Popular/i })).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: /Upcomming/i })).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: /Free Games/i })).toBeInTheDocument()
  })

  it('should render Banner, Highlight and GameCard', () => {
    makeSut()
    // GameCard + Banner
    expect(screen.getAllByText(/Red Dead Redemption 2/i)).toHaveLength(7)
    // highlight
    expect(screen.getAllByText(/Read Dead it's back/i)).toHaveLength(3)
  })
})
