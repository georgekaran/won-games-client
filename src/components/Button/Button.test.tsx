import React from 'react'
import { RenderResult, screen } from '@testing-library/react'

import Button, { ButtonProps } from './Button'
import { renderWithTheme } from '@/test/helpers'
import theme from '@/styles/theme'

type SutProps = Pick<ButtonProps, 'size' | 'fullWidth'>

const makeSut = ({ size = 'medium', fullWidth = false }: SutProps = {}): RenderResult => {
  return renderWithTheme(<Button size={size} fullWidth={fullWidth}>Button</Button>)
}

describe('<Button />', () => {
  test('should render with correct initial state', () => {
    const { container } = makeSut()
    expect(container.firstChild).toMatchSnapshot()
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '4rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.small)
    expect(button).toHaveStyleRule('padding', `${theme.spacings.xxsmall} ${theme.spacings.medium}`)
  })

  test('should render small if small size is provided', () => {
    const { container } = makeSut({ size: 'small' })
    expect(container.firstChild).toMatchSnapshot()
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '3rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.xsmall)
    expect(button).toHaveStyleRule('padding', theme.spacings.xxsmall)
  })

  test('should render large if large size is provided', () => {
    const { container } = makeSut({ size: 'large' })
    expect(container.firstChild).toMatchSnapshot()
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('height', '5rem')
    expect(button).toHaveStyleRule('font-size', theme.font.sizes.medium)
    expect(button).toHaveStyleRule('padding', `${theme.spacings.xxsmall} ${theme.spacings.xlarge}`)
  })

  test('should render fullWidth if prop is provided', () => {
    const { container } = makeSut({ fullWidth: true })
    expect(container.firstChild).toMatchSnapshot()
    const button = screen.getByRole('button', { name: /Button/ })
    expect(button).toHaveStyleRule('width', '100%')
  })
})
