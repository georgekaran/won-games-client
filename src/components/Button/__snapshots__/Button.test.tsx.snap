// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['<Button /> should render fullWidth if prop is provided 1'] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  width: 100%;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`

exports['<Button /> should render large if large size is provided 1'] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`

exports['<Button /> should render small if small size is provided 1'] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 3rem;
  font-size: 1.2rem;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`

exports['<Button /> should render with correct initial state 1'] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: none;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
>
  <span>
    Button
  </span>
</button>
`
